o1vm/src/mips/tests.rs

```diff
@@ -323,4 +335,200 @@ mod folding {
         interpret_itype(&mut dummy_env, ITypeInstruction::LoadUpperImmediate);
         assert_eq!(dummy_env.registers.general_purpose[1], 0xa0000);
     }
+
+    fn make_random_witness_for_addiu<RNG>(
+        domain_size: usize,
+        _rng: &mut RNG,
+    ) -> Witness<N_MIPS_REL_COLS, Vec<Fp>>
+    where
+        RNG: RngCore + CryptoRng,
+    {
+        let mut rng = o1_utils::tests::make_test_rng();
+        let mut dummy_env = dummy_env(&mut rng);
+        let instr = ITypeInstruction::AddImmediateUnsigned;
+        let a = std::array::from_fn(|_| Vec::with_capacity(domain_size));
+        let mut witness = Witness { cols: Box::new(a) };
+        // Building trace for AddImmediateUnsigned
+        (0..domain_size).for_each(|_i| {
+            // Registers should not conflict because RAMlookup does not support
+            // reading from/writing into the same register in the same
+            // instruction
+            let reg_src = rng.gen_range(0..10);
+            let reg_dest = rng.gen_range(10..20);
+            // we simulate always the same instruction, with some random values
+            // and registers
+            write_instruction(
+                &mut dummy_env,
+                InstructionParts {
+                    op_code: 0b001001,
+                    rs: reg_src,  // source register
+                    rt: reg_dest, // destination register
+                    // The rest is the immediate value
+                    rd: rng.gen(),
+                    shamt: rng.gen(),
+                    funct: rng.gen(),
+                },
+            );
+            interpret_itype(&mut dummy_env, instr);
+
+            for j in 0..SCRATCH_SIZE {
+                witness.cols[j].push(dummy_env.scratch_state[j]);
+            }
+            witness.cols[SCRATCH_SIZE].push(Fp::from(dummy_env.instruction_counter));
+            witness.cols[SCRATCH_SIZE + 1].push(Fp::from(0));
+
+            dummy_env.reset_scratch_state()
+        });
+        // sanity check
+        witness
+            .cols
+            .iter()
+            .for_each(|x| assert_eq!(x.len(), domain_size));
+        witness
+    }
+
+    fn build_folding_instance(
```

Comment:
Don't you prefer to implement `Foldable` for the normal `Trace`? I can adapt the signature so it matches better what you have in here.

Reply:
It would be nice to merge, yes. I was thinking we could do a better interface in the near future.